Change metadata properties to be placed in the JSON API meta object instead of using a suffix

rartino opened the pull request:
This tries to implement the end proposed result based on the discussion in today's web meeting.
I thought it was best to just get this down into text before we forgot.

Note that this is a PR against @JPBergsma's repo for the branch he is PR:ing to merge into the OPTIMADE devel branch.

This link gives a more direct comparison between OPTIMADE develop and the present proposed change:

https://github.com/Materials-Consortia/OPTIMADE/compare/develop...rartino:JPBergsma_add_meta

rartino committed:
Change metadata properties to be placed in the JSON API meta object instead of using a suffix

rartino commented:
@sauliusg, @vaitkus, @ml-evs, @merkys What do you think of this version of what we discussed today?

Note: go to this link and switch to the "Files changed (1)" tab to get the most clean view of what is suggested here:
https://github.com/Materials-Consortia/OPTIMADE/compare/develop...rartino:JPBergsma_add_meta

rartino committed:
Minor grammar corrections

rartino commented:
Allow me to ping this again to @JPBergsma, @sauliusg, @vaitkus, @ml-evs, @merkys, and also add @gmrigna. People seemed quite engaged in this discussion last meeting and expressed a level of urgency to resolve it; so I had hoped to see a bit of a faster turnaround on this concrete suggestion. Especially those of you with somewhat strong opinions, please make your feedback heard here to not hold up the process.

I re-iterate this is modification PR against @JPBergsma's PR for adding metadata properties. To see the whole proposed solution for metadata properties, use this link:

https://github.com/Materials-Consortia/OPTIMADE/compare/develop...rartino:JPBergsma_add_meta

and select the tab marked "Files changed (1)".

merkys commented:
I like this suggestion much better than prefixing/postfixing approach. However, the current specification already defined `meta` as "JSON API meta object that contains non-standard meta-information about the object", thus this definition has to be changed as well (how about backwards compatibility?).

rartino commented:
Good point - we should adjust that formulation as well - it is copied right out of the JSON:API specification; but if we standardize what goes in the field it is odd to also say that it is "non-standard". But, from the perspective of JSON:API and prior OPTIMADE specifications, I would argue that what we are standardizing here is "non-standard meta-information about the object" from their perspective.

merkys commented:
After re-reading JSON:API v1.0 specification I as well tend to think that "non-standard" is from their point of view. Furthermore, there do not seem to be any strict requirements concerning `meta`.

ml-evs commented on optimade.rst:
```diff
@@ -442,42 +442,62 @@ For example, the following query can be sent to API implementations `exmpl1` and
-This metadata field consists of a dictionary which MAY contain database specific fields.
-If an implementation supports the metadata field, it SHOULD return the metadata field whenever the property to which the metadata field belongs is returned.
+A metadata property represents property-specific metadata for each entry.
+How these are communicated in the response depend on the response format.
```
```suggestion
How these are communicated in the response depends on the response format.
```

ml-evs commented on optimade.rst:
```diff
@@ -442,42 +442,62 @@ For example, the following query can be sent to API implementations `exmpl1` and
 
-This metadata field consists of a dictionary which MAY contain database specific fields.
-If an implementation supports the metadata field, it SHOULD return the metadata field whenever the property to which the metadata field belongs is returned.
+A metadata property represents property-specific metadata for each entry.
```
```suggestion
A metadata property represents property-specific metadata for a given entry.
```

ml-evs commented on optimade.rst:
```diff
@@ -442,42 +442,62 @@ For example, the following query can be sent to API implementations `exmpl1` and
-If an implementation supports the metadata field, it SHOULD return the metadata field whenever the property to which the metadata field belongs is returned.
+A metadata property represents property-specific metadata for each entry.
+How these are communicated in the response depend on the response format.
+For the JSON response format, the resource object metadata field is used, see `JSON Response Schema: Common Fields`_.
```
```suggestion
For the JSON response format, the resource object metadata field, :field:`meta`, is used, see `JSON Response Schema: Common Fields`_.
```

ml-evs commented on optimade.rst:
```diff
@@ -442,42 +442,62 @@ For example, the following query can be sent to API implementations `exmpl1` and
 
-- **x-optimade-metadata-for**: This key contains a string with the name of the property for which this metadata property contains the metadata.
-  It MUST be located at the top level of the property definition.
+If an implementation supports the metadata field, it SHOULD return the metadata field whenever the property to which the metadata field belongs is returned.
```
How does this work in the case of one entry not having the associated metadata? e.g., adding an entry below to which there is no confidence interval, or it is invalid, e.g., a database that serves a theoretical structure with exact `elements_ratios: [1]` alongside experimental structures that have confidence intervals? Should the database be forced to assign a value (I guess confidence interval of 1.0 in this case), or are they allowed to omit the metadata field? Further down we say `meta` here is a `MAY` per resource but does that contradict the "SHOULD" here?

rartino replied:
The "SHOULD" here is meant to refer to including, e.g., the field `meta -> element_ratios` in which you would then put things like confidence intervals. The property definition for that meta field should specify whether the confidence interval is required or optional to include in that meta field. Depending on what the metadata property definition says, an empty `meta -> element_ratios` could be fine.

I don't have strong feelings how we should put the MAY and SHOULDs here, but the way it is presently is meant to convey that implementations MAY support metadata fields (but MAY thus chose not to). However, IF they support them, they SHOULD always include them in the `meta` part of the entry response if the corresponding data field is included in the response.

ml-evs reviewed (commented):
A few comments with suggestions below, but overall I also support this approach instead of the postfix/prefix.

rartino committed:
Minor grammar corrections from review

Co-authored-by: Matthew Evans <7916000+ml-evs@users.noreply.github.com>

rartino committed:
Fix old per-element meta field definition for metadata properties

rartino commented:
@merkys I think I've fixed the meta field definition in the place you were talking about (there are a *lot* of meta fields being defined, but as far as I could find, there were only one other place that refers to the per-entry meta field, which is the one I fixed.)

JPBergsma commented on optimade.rst:
```diff
@@ -442,42 +442,62 @@ For example, the following query can be sent to API implementations `exmpl1` and
-The subfields SHOULD have the same value for the :field:`$id` field if the :field:`$id` field is present and the subfields are otherwise identical.
-In addition to the fields described in the section `property definitions`_, the property definition of a metadata field SHOULD have the key:
+The metadata property is a dictionary of format specified by the field :field:`x-optimade-metadata-definition` in the Property Definition of the field, see `Property Definitions`_.
+Database providers are allowed to define their own metadata properties in :field:`x-optimade-metadata-definition`, but they MUST use the database-specific prefix even for metadata for database-specific fields.
```
```suggestion
Database providers are allowed to define their own metadata properties in :field:`x-optimade-metadata-definition`, but they MUST use the database-specific prefix in the .
```
When a database adds their own fields, they always have to be prefixed by the database specific prefix. So I do not see why we need to specify this here.

JPBergsma commented on optimade.rst:
```diff
@@ -442,42 +442,62 @@ For example, the following query can be sent to API implementations `exmpl1` and
-If a subfield is present in multiple metadata fields, these subfields should have a separate entry under each of these metadata fields.
-The subfields SHOULD have the same value for the :field:`$id` field if the :field:`$id` field is present and the subfields are otherwise identical.
-In addition to the fields described in the section `property definitions`_, the property definition of a metadata field SHOULD have the key:
+The metadata property is a dictionary of format specified by the field :field:`x-optimade-metadata-definition` in the Property Definition of the field, see `Property Definitions`_.
```
```suggestion
The metadata property is a dictionary in the format specified by the field :field:`x-optimade-metadata-definition` in the Property Definition of the field, see `Property Definitions`_.
```
For me "dictionary of format" sounds strange. I think "the" is missing.

JPBergsma commented on optimade.rst:
```diff
@@ -442,42 +442,62 @@ For example, the following query can be sent to API implementations `exmpl1` and
+     }
 
-Example of the property definition of a metadata field:
+Example of the corresponding metadata property definition to be placed in the field :field:`x-optimade-metadata-definition` in the property definition of :field:`element_ratios`:
```
This sounds a bit strange, as the example property definition for the metadata field below already contains the field `x-optimade-metadata-definition`
```suggestion
Example of the corresponding metadata property definition contained in the field :field:`x-optimade-metadata-definition` which is placed in the property definition of :field:`element_ratios`:
```

JPBergsma commented on optimade.rst:
```diff
@@ -1946,6 +1969,12 @@ A Property Definition MUST be composed according to the combination of the requi
+**OPTIONAL keys for the outermost level of the Property Definition:**
+
+- :field:`x-optimade-metadata-definition`: Dictionary.
+  A dictionary that is itself a property definition which defines the format of the metadata dictionary for this field.
```
```suggestion
  A dictionary that is itself a property definition, which defines the format of the metadata dictionary for this field.
```

JPBergsma commented:
I have added a few remarks.
One important point, is that this proposal does not support metadata having its own metadata. And therefore it is not possible to have ranged fields within metadata. The size of the metadata must therefore be quite small, as it is always provided.

I was also wondering whether nested fields are allowed to have metadata fields and how these should be represented.

I think the original intent of the JSON API specification is that this meta field contains meta data for the whole entry. 
How are we going to handle such meta data ? Could it be confusing there are fields under meta that do not correspond to one of the fields under attributes ?

rartino committed:
Minor language corrections from review

Co-authored-by: Johan Bergsma <29785380+JPBergsma@users.noreply.github.com>

rartino commented:
> One important point, is that this proposal does not support metadata having its own metadata. And therefore it is not possible to have ranged fields within metadata. The size of the metadata must therefore be quite small, as it is always provided. 

That is correct - and personally I'm more than happy to wait with support for these until we see motivated use cases. From a practical perspective, there will always be the option for a big metadata field to be turned into a data field.

> I was also wondering whether nested fields are allowed to have metadata fields and how these should be represented.

Nested fields in the data? There is one single metadata property definition that controls how the metadata is provided for one property. If that property is nested, then it is up to the one defining the metadata property definition to decide how things are organized inside the metadata field.

> I think the original intent of the JSON API specification is that this meta field contains meta data for the whole entry.
How are we going to handle such meta data ? Could it be confusing there are fields under meta that do not correspond to one of the fields under attributes ?

My view is that metadata relating to attributes of an entry is metadata that pertains to that entry. But if we in the future want to communicate "more global" entry metadata, I don't see why we couldn't reserve a name, e.g., `"entry": { ... }`, and place it there.

..

JPBergsma commented:
> That is correct - and personally I'm more than happy to wait with support for these until we see motivated use cases. From a practical perspective, there will always be the option for a big metadata field to be turned into a data field.

One possible use case that we came up with were the confidence intervals, which could be as large as the original property. I am mostly wondering how we could implement that given the current PR. It does not need to be in this PR, but we should have an idea of how it could reasonably be done.

 Could we perhaps make a structure like:
```
"meta":{
  "attributes":{ 
    "element_ratios":{
      "metadata":{ // This field would contain all the metadata belonging to the field one level above. 
          "_exmpl_confidence_interval": [[0.33325,0.33347],[0.22190,0.22268],[0.44392,0.44458]],
          "_exmpl_confidence_level": 0.95
      },
      //...
  },
  "metadata":{  
    "creator":"Some Name", // Here we could store the metadata that applies to the entry as a whole.
  },
  "meta":{
    // Here we could place meta data for the metadata fields themselves
  } 
}
```
The field `metadata`is special in this case in that it contains the actual metadata whereas all other fields refer to nested properties. 
It would solve a lot of the points mentioned before: Where metadata for the whole entry should go, how to store metadata for nested fields and how to store metadata for the fields in metadata itself.

JPBergsma commented:
I have added most of the changes you suggested here to the file. The difference we can discuss during the OPTIMADE meeting next week.
